src/stcal/dq_init/dq_initialization.py

```diff
@@ -0,0 +1,99 @@
+import logging
+import numpy as np
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+def do_dqinit(input_model, mask_model, guider_list=None):
+    """Perform the dq_init step on a datamodel
+
+    Parameters
+    ----------
+    input_model : input science datamodel
+        The science datamodel to be corrected
+
+    mask_model : mask datamodel
+        The mask model to use in the correction
+
+    Returns
+    -------
+    output_model : science datamodel
+        The dq corrected science datamodel
+    """
+
+    # Inflate empty DQ array, if necessary
+    check_dimensions(input_model)
+
+    # Create output model as copy of input
+    output_model = input_model.copy()
+
+    mask_array = mask_model.dq
+
+    # Set model-specific data quality in output
+    if input_model.meta.exposure.type in guider_list:
+        dq = np.bitwise_or(input_model.dq, mask_array)
+        output_model.dq = dq
+    else:
+        dq = np.bitwise_or(input_model.pixeldq, mask_array)
+        output_model.pixeldq = dq
+
+    output_model['meta']['cal_step']['dq_init'] = 'COMPLETE'
+
+    return output_model
+
+
+def check_dimensions(input_model):
+    """Check that the input model pixeldq attribute has the same dimensions as
+    the image plane of the input model science data
+    If it has dimensions (0,0), create an array of zeros with the same shape
+    as the image plane of the input model.
+
+    For the guiding modes, the GuiderRawModel has only a regular dq array (no pixeldq or groupdq)
+
+    Parameters
+    ----------
+    input_model : Raw datamodel
+        input datamodel
+
+    Returns
+    -------
+    None
+    """
+
+    input_shape = input_model.data.shape
+
+    # Temporarily commented until GuiderModels implemented
+    # if isinstance(input_model, datamodels.GuiderRawModel):
+    #     if input_model.dq.shape != input_shape[-2:]:
+    #
+    #         # If the shape is different, then the mask model should have
+    #         # a shape of (0,0).
+    #         # If that's the case, create the array
+    #         if input_model.dq.shape == (0, 0):
+    #             input_model.dq = np.zeros((input_shape[-2:])).astype('uint32')
+    #         else:
+    #             log.error("DQ array has the wrong shape: (%d, %d)" %
+    #                       input_model.dq.shape)
+    #
+    # else:   # RampModel
```

Comment:
If this is commented out how is it going to be used by JWST. 
Would it be better to leave `check_dimensions` in the respective pipelines?

Reply:
Guider data models were implemented a long time ago, so this code could be updated or activated on the JWST side. Obviously no one has looked at this step in such a long time that this to-do item was forgotten.

Reply:
Wait. I don't see this commented code in the current JWST implementation anywhere. Where did it come from? Was this commented out for the purpose of the Roman implementation?

Reply:
Yeah, it's commented out here, but not in `jwst`.  So that's a non-starter.

Howard, is there a better way we can identify this data other than using an `isinstance()` call?  I.e. is there a specific EXP_TYPE or similar that could live here and would not need access to `jwst.datamodels.GuiderRawModel`?  I.e. how do we know when to run `calwebb_guider` on an exposure?

Reply:
Yes, the jwst version defines a list of EXP_TYPE's that are exclusive to guide star data, so any datamodel that has one of those EXP_TYPE's should be a straight substitute for testing the model instance. For JWST the selection of `calwebb_guider` is done within the SDP workflow manager. But the reason for this test in the `dq_init` step is because that step is used for both normal science processing (`calwebb_detector1`) and guide star data processing (`calwebb_guider`). And unfortunately the content of the science and guider data models is slightly different, in that science modes have PIXELDQ and GROUPDQ arrays, while guider only has a generic DQ array. Hence the need for this test in the `dq_init` step.